src/mono/mono/mini/aot-compiler.c

```diff
@@ -6111,6 +6114,90 @@ method_is_externally_callable (MonoAotCompile *acfg, MonoMethod *method)
 	}
 }
 
+#ifdef MONO_ARCH_AOT_SUPPORTED
+static gboolean
+method_to_pinvoke_has_scope_import (MonoAotCompile *acfg, MonoMethod *method, const char **module, const char **entrypoint)
+{
+	MonoImage *image = m_class_get_image (method->klass);
+	MonoMethodPInvoke *piinfo = (MonoMethodPInvoke *) method;
+	MonoTableInfo *tables = image->tables;
+	MonoTableInfo *im = &tables [MONO_TABLE_IMPLMAP];
+	MonoTableInfo *mr = &tables [MONO_TABLE_MODULEREF];
+	guint32 im_cols [MONO_IMPLMAP_SIZE];
+	int module_idx;
+	const char **scope_import;
+	guint32 scope_token;
+
+	if (g_hash_table_lookup_extended (acfg->method_to_pinvoke_scope_import, method, NULL, (gpointer *)&scope_import) && scope_import) {
+		if (module)
+			*module = g_strdup (scope_import[0]);
+		if (entrypoint)
+			*entrypoint = g_strdup (scope_import[1]);
+		return TRUE;
+	}
+
+	if (piinfo->implmap_idx == 0 || mono_metadata_table_bounds_check (image, MONO_TABLE_IMPLMAP, piinfo->implmap_idx))
+		return FALSE;
+
+	mono_metadata_decode_row (im, piinfo->implmap_idx - 1, im_cols, MONO_IMPLMAP_SIZE);
+
+	module_idx = im_cols [MONO_IMPLMAP_SCOPE];
+	if (module_idx == 0 || mono_metadata_table_bounds_check (image, MONO_TABLE_MODULEREF, module_idx))
+		return FALSE;
+
+	scope_import = (const char **) g_malloc0 (3 * sizeof (scope_import));
```

Comment:
why do you need 3 slots? Is it because you use `g_strfreev` that expects NULL? I think it is better to save 4/8 bytes and implement a simple custom free function for this particular scenario like:

```
static void
free_method_pinvoke_import_value (void *data)
{
    char **value = (char **)data;
    if (!value)
        return;
    _free (value[0]);
    _free (value[1]);
}
```

and then use that as hash table free function.

Since each slot is of type char * you can use sizeof(char *)

scope import should be char ** instead of const char **

Reply:
Right, I had noticed that `g_strfreev` is expected to be a NULL terminated string of arrays. At what point would it be better to use `g_strfreev` instead of making a custom free function? Is it because there are only 2 strings in the array that we wouldnt want to allocate +50% memory just to hold NULL?

How come scope import shouldn't be `const char **`? Doesn't it point to an address that points to a string that is expected to not change (the scope and import value strings)?

Reply:
Regarding `g_strfreev`, there is no hard rules and it could be argued both ways, but if you deal with smaller arrays with known size, reserving one slot of size pointer to hold NULL, just to be able to use `g_strfreev`, feels like a waste that can add up if you have many items in the table.

Regarding `const char **`, in this case its an array to strings that gets passed into hash table that will free them on destruction, so that means data is modifiable until added to hash table, but when you extract out a copy of it form the hash table you should handle it as const char ** or even const char * const * since both array as well as each string should be considered read only.